10337 current cryptography

glyph opened the pull request:
<!--
## Remove this section

Have a look at [our developer documentation](https://twistedmatrix.com/trac/wiki/TwistedDevelopment#SubmittingaPatch) before submitting your Pull Request.

Note that the Trac ticket, news fragment, and review submission portions of this process apply to *all* pull requests, no matter how small;
if you don't do them, it's likely that nobody will even notice your PR needs a review.
-->

## Scope and purpose

Add a few words about why this PR is needed and what is its scope.

Add any comments about trade-offs (if any) made in this PR and the reasoning behind them.

Add mentions of things that are not covered here and are planed to be done in separate PRs.


## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/10337
* [x] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [x] The title of the PR starts with the associated Trac ticket number (without the `#` character).
* [x] I have updated the automated tests and checked that all checks for the PR are green.
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.
* [x] The merge commit will use the below format
  The first line is automatically generated by GitHub based on PR ID and branch name.
  The other lines generated by GitHub should be replaced.

```
Merge pull request #123 from twisted/4356-branch-name-with-trac-id

Author: glyph
Reviewer: adiroiban
Fixes: ticket: 10337

Twisted now supports Cryptography version 37 and later.
```

glyph committed:
regenerate server.pem, upgrade TLS method for testing

glyph committed:
while we're in here, no <2048 RSA please

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

alex commented on src/twisted/protocols/test/test_tls.py:
```diff
@@ -72,11 +74,7 @@ class HandshakeCallbackContextFactory:
-    SSL_CB_HANDSHAKE_DONE = 0x20
-
-    def __init__(self, method=TLSv1_METHOD):
+    def __init__(self, method=TLS_METHOD):
```
Minimum pyOpenSSL of [21.0.0](https://github.com/pyca/pyopenssl/releases/tag/21.0.0)

glyph committed:
remove tons of deprecated / dangerous TLS version references

glyph committed:
adjust for defaults

glyph committed:
preserve behavior when cryptography is not installed

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

glyph committed:
news fragment

glyph committed:
Merge branch 'trunk' into 10337-current-cryptography

glyph committed:
remove unused imports

glyph committed:
import handshake done constant

glyph committed:
isort

glyph committed:
cryptography minimum version (thanks @alex)

glyph committed:
turns out this never did anything anyway, so we can kill it

glyph committed:
minor detail fixes

glyph committed:
fix error translation for new pyOpenSSL version

glyph committed:
shape the error queue correctly

glyph committed:
renegotiation is a bad idea, remove this failing test

OpenSSL 3 seems to introduce some new hurdles to renegotiation, but you
shouldn't be doing it anyway; it only works if you restrict yourself to
TLSv1.2.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

glyph committed:
generalize represents-eof detection

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

glyph committed:
remove unused import

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

glyph committed:
fix small coverage gap

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

adiroiban reviewed (approved):
Much appreciated. Great work! Looks good and I think it should be merged.
